cdap-app-fabric/src/main/java/io/cdap/cdap/app/store/ApplicationFilter.java

```diff
@@ -0,0 +1,92 @@
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.app.store;
+
+import io.cdap.cdap.api.artifact.ArtifactId;
+import io.cdap.cdap.proto.id.ApplicationId;
+
+import java.util.Locale;
+import java.util.Set;
+import java.util.function.Predicate;
+
+/**
+ * This class defined various filters that can be applied during applicaiton store scanning. The filters itself
+ * are pushed down, so only the ones defined in this class and it's inner classes are usually supported.
+ * Attempt to use any other will result in an {@link UnsupportedOperationException}. Different filter types
+ * analyze different attributes and operate on those attributes without full information retrieval / deserialization.
+ */
+public abstract class ApplicationFilter {
+
+  /**
+   * Base class for filters that filter on {@link ApplicationId} data. This are the fastest filter as they don't
+   * need full application record.
+   */
+  public abstract static class ApplicationIdFilter extends ApplicationFilter implements Predicate<ApplicationId> { }
+
+  /**
+   * Base class for filters that filter on application {@link ArtifactId}.
+   */
+  public abstract static class ArtifactIdFilter extends ApplicationFilter implements Predicate<ArtifactId> { }
+
+  /**
+   * The filter that check if application id contains a substring (ignoring case).
+   */
+  public static class ApplicationIdContains extends ApplicationIdFilter {
+    private final String searchFor;
+
+    public ApplicationIdContains(String searchFor) {
+      this.searchFor = searchFor.toLowerCase(Locale.ROOT);
+    }
+
+    @Override
+    public boolean test(ApplicationId applicationId) {
+      return applicationId.getApplication().toLowerCase(Locale.ROOT).contains(searchFor);
+    }
+  }
+
+  /**
+   * Returns true if the application artifact is in a allow list of names
+   */
+  public static class ArtifactNamesIn extends ArtifactIdFilter {
+    private final Set<String> names;
+
+    public ArtifactNamesIn(Set<String> names) {
+      this.names = names;
+    }
+
+    @Override
+    public boolean test(ArtifactId input) {
+      return names.contains(input.getName());
+    }
+  }
+
+  /**
+   * Returns true if the application artifact is a specific version
+   */
+  public static class ArtifactVersionPredicate extends ArtifactIdFilter {
+    private final String version;
+
+    public ArtifactVersionPredicate(String version) {
+      this.version = version;
+    }
+
+    @Override
+    public boolean test(ArtifactId input) {
+      return version.equals(input.getVersion().getVersion());
```

Comment:
wanna handle the case when version is null ?

Reply:
No, the version is non-null required parameter. If anyone pass null, it's better to fail as it's coding error